integration/data/test_snapshot.py

```diff
@@ -366,3 +371,96 @@ def test_expansion_with_backing_file(grpc_backing_controller,  # NOQA
+    mount_cmd = nsenter_cmd + ["mount", "--make-shared", dev_path, mnt_path]
+    umount_cmd = nsenter_cmd + ["umount", mnt_path]
+    findmnt_cmd = nsenter_cmd + ["findmnt", dev_path]
+    subprocess.check_call(nsenter_cmd + ["mkfs.ext4", dev_path])
+    subprocess.check_call(nsenter_cmd + ["mkdir", "-p", mnt_path])
+    subprocess.check_call(mount_cmd)
+    subprocess.check_call(findmnt_cmd)
+
+    def checksum_test_file():
+        read_cmd = nsenter_cmd + ["cat", test_file]
+        data = subprocess.check_output(read_cmd)
+        return checksum_data(str(data).encode('utf-8'))
+
+    def write_test_file():
+        # beware don't touch this write command
+        data = random_string(length)
+        write_cmd = ["/bin/sh -c '/bin/echo",
+                     '"' + data + '"', ">", test_file + "'"]
+        shell_cmd = " ".join(nsenter_cmd + write_cmd)
+
+        subprocess.check_call(shell_cmd, shell=True)
+        return checksum_test_file()
+
+    # create snapshot1 with empty fs
+    # NOTE: we cannot use checksum_dev since it assumes
+    #  asci data for device data instead of raw bytes
+    snap1 = cmd.snapshot_create(address)
+
+    # create snapshot2 with a new test file
+    test2_checksum = write_test_file()
+    snap2 = cmd.snapshot_create(address)
+
+    # create snapshot3 overwriting the test file
+    test3_checksum = write_test_file()
+    snap3 = cmd.snapshot_create(address)
+
+    # verify existence of the snapshots
+    snapshots = cmd.snapshot_ls(address)
+    assert snap1 in snapshots
+    assert snap2 in snapshots
+    assert snap3 in snapshots
+
+    # unmount the volume, since each revert will shutdown the device
+    subprocess.check_call(umount_cmd)
+
+    # restore snapshots 1,2,3 & verify filesystem state
+    print("\nsnapshot_revert_with_frontend snap1 begin")
+    snapshot_revert_with_frontend(address, engine_name, snap1)
+    print("snapshot_revert_with_frontend snap1 finish\n")
+    subprocess.check_call(mount_cmd)
+    # should error since the file does not exist in snapshot 1
+    with pytest.raises(subprocess.CalledProcessError):
+        print("is expected error, since the file does not exist.")
+        checksum_test_file()
+    subprocess.check_call(umount_cmd)
+
+    print("\nsnapshot_revert_with_frontend snap2 begin")
+    snapshot_revert_with_frontend(address, engine_name, snap2)
+    print("snapshot_revert_with_frontend snap2 finish\n")
+    subprocess.check_call(mount_cmd)
+    assert checksum_test_file() == test2_checksum
+    subprocess.check_call(umount_cmd)
+
+    print("\nsnapshot_revert_with_frontend snap3 begin")
+    snapshot_revert_with_frontend(address, engine_name, snap3)
+    print("snapshot_revert_with_frontend snap3 finish\n")
+    subprocess.check_call(mount_cmd)
+    assert checksum_test_file() == test3_checksum
+    subprocess.check_call(umount_cmd)
+
+    # remove the created mount folder
+    subprocess.check_call(nsenter_cmd + ["rmdir", mnt_path])
```

Comment:
The mount directory is not empty. I think this doesn't work

Reply:
The volume is unmounted as such the folder used for the mount point does not contain any data?

Reply:
Oh, sorry.